### PR TITLE
Allow database migration upon adding new database column family to schema

### PR DESCRIPTION
**Pull request by simrankedia**
Adds migration capabilities (allows adding new column family to existing database) to RocksDB instance. This change makes sure RocksDB open is always requested with the existing set of column families and all new column families are later created once opened.

- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests - Tests introduced by https://github.com/FuelLabs/fuel-core/pull/1853

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

**Comment by xgreenx** [comment]
Answering your comment in the issue:

> @Voxelot I have tried to raise a PR to fix this issue #1851 - please let me know if the direction is correct - I will accordingly complete and release a new PR with tests etc.

You are welcome to contribute, and yeah, it is the right direction.

I created a test case https://github.com/FuelLabs/fuel-core/pull/1853 to prove that the fix works. Could you rebase your PR to use this branch with the test case, please?

File: crates/fuel-core/src/state/rocks_db.rs [hunk 1]
```diff
@@ -227,34 +220,48 @@ where
...
-                }
+        let existing_column_families = DB::list_cf(&opts, &path).map_err(|err| {
+            tracing::error!("Couldn't get the list of cfs: {}.\n", err);
+            DatabaseError::Other(err.into())
+        })?;
+
+        let mut cf_descriptors_to_open = vec![];
+        let mut cf_descriptors_to_create= vec![];
+        for column in columns.clone() {
+            let column_name = format!("col-{}", column.id());
```
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:231]:
```suggestion
            let column_name = Self::col_name(column.id());
```

File: crates/fuel-core/src/state/rocks_db.rs [hunk 2]
```diff
@@ -227,34 +220,48 @@ where
...
+            if existing_column_families.contains(&column_name) {
+                cf_descriptors_to_open.push(ColumnFamilyDescriptor::new(
+                    Self::col_name(column.id()),
```
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:234]:
```suggestion
                    column_name,
```

File: crates/fuel-core/src/state/rocks_db.rs [hunk 3]
```diff
@@ -227,34 +220,48 @@ where
...
+                    Self::cf_opts(column, &block_opts),
+                ));
+            } else {
+                let opts = Self::cf_opts(column, &block_opts);
+                cf_descriptors_to_create.push((Self::col_name(column.id()), opts));
```
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:239]:
```suggestion
                cf_descriptors_to_create.push((column_name, opts));
```

**Review by xgreenx — COMMENTED** [review]
Could you fix CI please?=) You can run locally `./ci_checks.sh` to perform all CI checks

**Comment by simrankedia** [comment]
> Could you fix CI please?=) You can run locally `./ci_checks.sh` to perform all CI checks

@xgreenx  I am seeing the following issue upon running the checks, which seems unrelated. 
Screenshot 2024-04-23 at 00 20 23

**Comment by simrankedia** [comment]
> Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/FuelLabs/fuel-core?pullRequest=1851) before we can accept your contribution.
> **1** out of **2** committers have signed the CLA.

@xgreenx  I think there has been some messups given the git config setup locally on my system. I am not sure how to get rid of it. I will create a duplicate PR getting rid of this extra contributor.

**Comment by xgreenx** [comment]
> I am seeing the following issue upon running the checks, which seems unrelated.

I would suggest you to use `rust 1.75`=)

**Comment by simrankedia** [comment]
> The change fails most of tests=) 
> 
> image
> 
> 

Yes, I will look into it and get back. Didn't get a chance today.

**Comment by simrankedia** [comment]
> request as draft [ April 24, 2024 14:14 ](#event-12591782371)

I have fixed the original issue and the failing tests have been fixed. However, upon further running the checks I see some 5 more tests failing: 

```
    poa::p2p::test_poa_multiple_producers
    sync::test_multiple_producers_different_keys
    sync::test_partitions_larger_groups::partition_with_100_txs_8_validators_4_partitions
    sync::test_partitions_larger_groups::partition_with_10_txs_8_validators_4_partitions
    sync::test_partitions_larger_groups::partition_with_1_tx_3_validators_3_partitions
```

These are not directly related to my changes atleast - I will investigate more tomorrow. Let me know if you have any idea till then. Thanks

**Comment by simrankedia** [comment]
> > request as draft [ April 24, 2024 14:14 ](#event-12591782371)
> 
> I have fixed the original issue and the failing tests have been fixed. However, upon further running the checks I see some 5 more tests failing:
> 
> ```
>     poa::p2p::test_poa_multiple_producers
>     sync::test_multiple_producers_different_keys
>     sync::test_partitions_larger_groups::partition_with_100_txs_8_validators_4_partitions
>     sync::test_partitions_larger_groups::partition_with_10_txs_8_validators_4_partitions
>     sync::test_partitions_larger_groups::partition_with_1_tx_3_validators_3_partitions
> ```
> 
> These are not directly related to my changes atleast - I will investigate more tomorrow. Let me know if you have any idea till then. Thanks

@xgreenx  Slightly by increasing this `time_limit` (https://github.com/FuelLabs/fuel-core/blob/master/crates/fuel-core/src/p2p_test_helpers.rs#L383) all test passes - I am running an experiment to understand what kind of limit to use - but wanted to check if there are any specific constraints that needs to be followed here?

File: crates/fuel-core/src/state/rocks_db.rs [hunk 4]
```diff
@@ -227,34 +220,53 @@ where
...
+            }
+        }
+        let db = match DB::open_cf_descriptors(&opts, &path, cf_descriptors_to_open) {
+            Ok(db) => {
+                // Setup cfs
+                for (name, opt) in cf_descriptors_to_create {
+                    db.create_cf(name, &opt)
+                        .map_err(|e| DatabaseError::Other(e.into()))?;
+                }
+                Ok(db)
```
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:255]: Maybe it makes sense to move this loop after `db` is opened(below of the `match` statement). In this case you also will handle the case of the error with repair.

File: crates/fuel-core/src/state/rocks_db.rs [hunk 5]
```diff
@@ -227,34 +220,53 @@ where
...
+                DB::repair(&opts, &path)
+                    .map_err(|e| DatabaseError::Other(e.into()))?;
+
+                let cf_descriptors = columns.clone().into_iter().map(|i| {
+                    ColumnFamilyDescriptor::new(
+                        Self::col_name(i.id()),
+                        Self::cf_opts(i, &block_opts),
+                    )
+                });
+                DB::open_cf_descriptors(&opts, &path, cf_descriptors)
```
Comment (xgreenx) [inline crates/fuel-core/src/state/rocks_db.rs:268]:
```suggestion
                DB::open_cf_descriptors(&opts, &path, cf_descriptors_to_open)
```

**Pull request merged**